### PR TITLE
minimal-bootstrap.live-bootstrap: add helper for fetching useful files

### PR DESCRIPTION
**Pull request by 06kellyjac**
###### Description of changes

The live-bootstrap project has a handful of files that make building this
software easier during the bootstrapping process. Generally these will be
Makefiles or patches.

This change adds a script to populate hashes in sources.nix and some
helper nix to make it easy to reference the files we want to fetch.

The fetcher script can fetch hashes in parallel and will skip already
fetched hashes unless told to fetch all so finishes almost instantly.

This method can be used at least until we can fetch a full live-bootstrap
bundle. Fetching the live-bootstrap tar.gz and unpacking it early in the
bootstrapping process is difficult due to 1. symlinks (not too much of an issue)
and 2. the very long filenames due to `<project name>-<full sha>` being the parent dir.

I've tried a couple different options but this felt like the best UX.
An alternative would be vendoring the files from live-bootstrap.
Fetching these individual files by hand is pretty painful though so either way
I think this is an improvement.

---

```text
λ ./pkgs/os-specific/linux/minimal-bootstrap/live-bootstrap/fetch-sources.py -h
usage: fetch-sources.py [-h] [-r | --regen | --no-regen] [-o | --overwrite | --no-overwrite] [-i [INPUT]]

options:
  -h, --help            show this help message and exit
  -r, --regen, --no-regen
                        Regenerate all hashes
  -o, --overwrite, --no-overwrite
                        Overwrite sources.nix relative to the script
  -i [INPUT], --input [INPUT]
                        Input sources.nix
```

---

I'm open to reformatting the python or squashing some of the code together so it's less verbose.
I don't write much python but I thought this would be better recieved than nodejs

Since `live-bootstrap` isn't a package itself but a set of helpers this might need some redesign.
Maybe going into utils or something?...

cc: @emilytrau @Ericson2314

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [X] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by 06kellyjac** [comment]
There are a couple of patches that I've seen on your other branch @emilytrau but don't seem to be here:

* `sysa/gzip-1.2.4/files/stat_override.c`
* `sysa/tar-1.12/files/getdate_stub.c`
* `sysa/tar-1.12/files/stat_override.c`

Should we adopt those now? Or are they unnecessary?

**Comment by emilytrau** [comment]
Those patches weren't necessary as we can now let ./configure sort most of it out. Shouldn't need to pull in makefiles from live-bootstrap from here on out!

File: pkgs/os-specific/linux/minimal-bootstrap/live-bootstrap/fetch-sources.py [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+  output_dict = json.loads(output)
+  regen = args.regen
+  if regen:
+    log("updating all hashes")
+  else:
+    log("updating only missing hashes")
+  # start concurrently fetching sources
+  new_sources = fetch_sources(regen, output_dict)
+  # squash together the old sources with the updates
+  merged_sources = update_sources(output_dict, new_sources)
```
Comment (06kellyjac) [inline pkgs/os-specific/linux/minimal-bootstrap/live-bootstrap/fetch-sources.py:144]: probably don't need to bother merging/updating when `regen = True`...

**Comment by Artturin** [comment]
Is this still relevant/necessary, if so please rebase.

@emilytrau you should apply for a commit bit in issue 50105.

**Comment by 06kellyjac** [comment]
I can trim down the list to what we currently need if the others think its valuable.

I think its quite clean and makes the names nicer in a big dependency graph

**Comment by emilytrau** [comment]
I'm so sorry the code is good but I'm not convinced this is a necessary abstraction to add for the small amount of live-bootstrap patches we use. I love the intention and effort put into the UX but I don't think it's the right fit for the size of the problem :'(

I think this would have been a great UX for `bootstrap-sources` if we didn't have the `make-bootstrap-sources`-in-cache solution

**Pull request closed without merging**